Revoking client role from group does not work

If I try to remove a client role from a group, I get an error:

```
[ERROR] 2024-07-08 15:48:08.736 [main] KeycloakMigration - Error occurred while migrating: clean-deprecated-roles 
de.klg71.keycloakmigration.changeControl.actions.MigrationException: Group with name: GROUP_A in realm: testRealm does not have role: DELETE_USER!
        at de.klg71.keycloakmigration.changeControl.actions.group.RevokeRoleFromGroupAction.execute(RevokeRoleFromGroupAction.kt:41) ~[keycloakmigration-0.2.56.RC3-fat.jar:?]
        at de.klg71.keycloakmigration.changeControl.actions.Action.executeIt(Action.kt:37) ~[keycloakmigration-0.2.56.RC3-fat.jar:?]
```

```
id: remove-deprecated-roles
author: mfolnovic
changes:
  - revokeRoleFromGroup:
      realm: testRealm
      role: DELETE_USER
      group: GROUP_A
      clientId: test-client
```

--- a/keycloakapi/src/main/kotlin/de/klg71/keycloakmigration/keycloakapi/KeycloakClient.kt
+++ b/keycloakapi/src/main/kotlin/de/klg71/keycloakmigration/keycloakapi/KeycloakClient.kt
@@ -424,6 +424,10 @@ interface KeycloakClient {
     @RequestLine("GET /admin/realms/{realm}/groups/{group-id}/role-mappings/realm/composite")
     fun groupRoles(@Param("realm") realm: String, @Param("group-id") id: UUID): List<RoleListItem>
 
+    @RequestLine("GET /admin/realms/{realm}/groups/{group-id}/role-mappings/clients/{client-id}/composite")
+    fun groupClientRoles(@Param("realm") realm: String, @Param("group-id") id: UUID,
+         @Param("client-id") clientId: UUID): List<RoleListItem>
+
     @RequestLine("GET /admin/realms/{realm}/clients/{client-id}/scope-mappings/realm")
     fun realmRoleScopeMappingsOfClient(@Param("realm") realm: String, @Param("client-id") id: UUID): List<RoleListItem>
 

--- a/src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/group/RevokeRoleFromGroupAction.kt
+++ b/src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/group/RevokeRoleFromGroupAction.kt
@@ -36,9 +36,20 @@ class RevokeRoleFromGroupAction(
             }
         }
 
-        client.groupRoles(realm(), client.groupUUID(group, realm())).run {
-            if (!map { it.name }.contains(role)) {
-                throw MigrationException("Group with name: $group in realm: ${realm()} does not have role: $role!")
+        if (clientId != null) {
+            client.groupClientRoles(
+                realm(), client.groupUUID(group, realm()), client.clientUUID(clientId, realm())
+            ).run {
+                if (!map { it.name }.contains(role)) {
+                    throw MigrationException(
+                        "Group with name: $group in client: $clientId in realm: ${realm()} does not have role: $role!")
+                }
+            }
+        } else {
+            client.groupRoles(realm(), client.groupUUID(group, realm())).run {
+                if (!map { it.name }.contains(role)) {
+                    throw MigrationException("Group with name: $group in realm: ${realm()} does not have role: $role!")
+                }
             }
         }
 

--- a/src/test/kotlin/de/klg71/keycloakmigration/changeControl/actions/group/RevokeRoleFromGroupIntegTest.kt
+++ b/src/test/kotlin/de/klg71/keycloakmigration/changeControl/actions/group/RevokeRoleFromGroupIntegTest.kt
@@ -3,9 +3,9 @@ package de.klg71.keycloakmigration.changeControl.actions.group
 import de.klg71.keycloakmigration.AbstractIntegrationTest
 import de.klg71.keycloakmigration.changeControl.actions.MigrationException
 import de.klg71.keycloakmigration.changeControl.actions.client.AddSimpleClientAction
-import de.klg71.keycloakmigration.changeControl.actions.client.UpdateClientAction
 import de.klg71.keycloakmigration.changeControl.actions.role.AddRoleAction
 import de.klg71.keycloakmigration.keycloakapi.KeycloakClient
+import de.klg71.keycloakmigration.keycloakapi.clientById
 import de.klg71.keycloakmigration.keycloakapi.groupByName
 import de.klg71.keycloakmigration.keycloakapi.model.RoleListItem
 import org.assertj.core.api.Assertions.assertThat
@@ -32,6 +32,26 @@ class RevokeRoleFromGroupIntegTest : AbstractIntegrationTest() {
         val role = RoleListItem(UUID.randomUUID(), role, null, false, false, testRealm)
 
         client.groupRoles(testRealm, client.groupByName(group, testRealm).id).let {
+            assertThat(it).usingElementComparatorOnFields("name", "containerId").doesNotContain(role)
+        }
+    }
+
+    @Test
+    fun testRevokeClientRole() {
+        AddGroupAction(testRealm, group).executeIt()
+        AddSimpleClientAction(testRealm, clientId).executeIt()
+        AddRoleAction(testRealm, role, clientId).executeIt()
+        AssignRoleToGroupAction(testRealm, role, group, clientId).executeIt()
+        RevokeRoleFromGroupAction(testRealm, role, group, clientId).executeIt()
+
+        val role = RoleListItem(UUID.randomUUID(), role, null, false, false,
+            client.clientById(clientId, testRealm).id.toString())
+
+        client.groupClientRoles(
+            testRealm,
+            client.groupByName(group, testRealm).id,
+            client.clientById(clientId, testRealm).id
+        ).let {
             assertThat(it).usingElementComparatorOnFields("name", "containerId").doesNotContain(role)
         }
     }
@@ -73,5 +93,16 @@ class RevokeRoleFromGroupIntegTest : AbstractIntegrationTest() {
             RevokeRoleFromGroupAction(testRealm, role, group).executeIt()
         }.isInstanceOf(MigrationException::class.java)
             .hasMessage("Group with name: $group in realm: $testRealm does not have role: $role!")
+    }
+
+    @Test
+    fun testAssignRole_clientRoleNotAssigned() {
+        AddGroupAction(testRealm, group).executeIt()
+        AddSimpleClientAction(testRealm, clientId).executeIt()
+        AddRoleAction(testRealm, role, clientId).executeIt()
+        assertThatThrownBy {
+            RevokeRoleFromGroupAction(testRealm, role, group, clientId).executeIt()
+        }.isInstanceOf(MigrationException::class.java)
+            .hasMessage("Group with name: $group in client: $clientId in realm: $testRealm does not have role: $role!")
     }
 }

